From: Padmanabh Ratnakar <padmanabh.ratnakar@emulex.com>
Date: Sat, 20 Oct 2012 06:02:40 +0000
Subject: [PATCH 121/143] be2net: Fix configuring VLAN for VF for Lancer
Patch-mainline: v3.8-rc1
Git-commit: a85e998681624ac85d423f34e833f95dffdd2a3c
References: bnc#777565 FATE#313819

Allow adding VLANs for Lancer VF.
VLAN ID 0 should not be added to list of VLANs sent to FW.

Signed-off-by: Padmanabh Ratnakar <padmanabh.ratnakar@emulex.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Benjamin Poirier <bpoirier@suse.de>
---
 drivers/net/ethernet/emulex/benet/be_main.c |   13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

--- a/drivers/net/ethernet/emulex/benet/be_main.c
+++ b/drivers/net/ethernet/emulex/benet/be_main.c
@@ -879,7 +879,12 @@ static void be_vlan_add_vid(struct net_d
 	struct be_adapter *adapter = netdev_priv(netdev);
 
 	adapter->vlans_added++;
-	if (!be_physfn(adapter))
+
+	if (!lancer_chip(adapter) && !be_physfn(adapter))
+		return;
+
+	/* Packets with VID 0 are always received by Lancer by default */
+	if (lancer_chip(adapter) && vid == 0)
 		return;
 
 	adapter->vlan_tag[vid] = 1;
@@ -893,7 +898,11 @@ static void be_vlan_rem_vid(struct net_d
 
 	adapter->vlans_added--;
 
-	if (!be_physfn(adapter))
+	if (!lancer_chip(adapter) && !be_physfn(adapter))
+		return;
+
+	/* Packets with VID 0 are always received by Lancer by default */
+	if (lancer_chip(adapter) && vid == 0)
 		return;
 
 	adapter->vlan_tag[vid] = 0;
